### PR TITLE
GH-101520: Move tracemalloc functionality into core, leaving interface in Modules.

### PR DESCRIPTION
**Pull request by markshannon**
This is a dumb refactoring.
All the new "API" functions are internal. They are marked `PyAPI_FUNC` for linkage.

* Issue: gh-101520

### CODE REVIEW

File: Include/tracemalloc.h [hunk 1]
```diff
@@ -33,6 +33,40 @@ PyAPI_FUNC(int) PyTraceMalloc_Untrack(
 PyAPI_FUNC(PyObject*) _PyTraceMalloc_GetTraceback(
     unsigned int domain,
     uintptr_t ptr);
+
+/* Return non-zero if traceamlloc is tracing */
```
Comment (Eclips4) [inline Include/tracemalloc.h]:
```suggestion
/* Return non-zero if tracemalloc is tracing */
```

File: Include/tracemalloc.h [hunk 2]
```diff
@@ -33,6 +33,40 @@ PyAPI_FUNC(int) PyTraceMalloc_Untrack(
...
+PyAPI_FUNC(int) _PyTraceMalloc_IsTracing(void);
+
+/* Clear the traceamlloc tracies */
```
Comment (Eclips4) [inline Include/tracemalloc.h]:
```suggestion
/* Clear the tracemalloc tracies */
```

File: Include/tracemalloc.h [hunk 3]
```diff
@@ -33,6 +33,40 @@ PyAPI_FUNC(int) PyTraceMalloc_Untrack(
...
+PyAPI_FUNC(void) _PyTraceMalloc_ClearTraces(void);
+
+/* Clear the traceamlloc tracies */
```
Comment (Eclips4) [inline Include/tracemalloc.h]:
```suggestion
/* Clear the tracemalloc tracies */
```

File: Include/tracemalloc.h [hunk 4]
```diff
@@ -33,6 +33,40 @@ PyAPI_FUNC(int) PyTraceMalloc_Untrack(
...
+PyAPI_FUNC(PyObject *) _PyTraceMalloc_GetTraces(void);
+
+/* Clear traceamlloc traceback for an object */
```
Comment (Eclips4) [inline Include/tracemalloc.h]:
```suggestion
/* Clear tracemalloc traceback for an object */
```

File: Include/tracemalloc.h [hunk 5]
```diff
@@ -33,6 +33,40 @@ PyAPI_FUNC(int) PyTraceMalloc_Untrack(
...
+PyAPI_FUNC(PyObject *) _PyTraceMalloc_GetObjectTraceback(PyObject *obj);
+
+/* Initialize traceamlloc */
```
Comment (Eclips4) [inline Include/tracemalloc.h]:
```suggestion
/* Initialize tracemalloc */
```

File: Include/tracemalloc.h [hunk 6]
```diff
@@ -33,6 +33,40 @@ PyAPI_FUNC(int) PyTraceMalloc_Untrack(
...
+PyAPI_FUNC(int) _PyTraceMalloc_Init(void);
+
+/* Start traceamlloc */
```
Comment (Eclips4) [inline Include/tracemalloc.h]:
```suggestion
/* Start tracemalloc */
```

File: Include/tracemalloc.h [hunk 7]
```diff
@@ -33,6 +33,40 @@ PyAPI_FUNC(int) PyTraceMalloc_Untrack(
...
+PyAPI_FUNC(int) _PyTraceMalloc_Start(int max_nframe);
+
+/* Stop traceamlloc */
```
Comment (Eclips4) [inline Include/tracemalloc.h]:
```suggestion
/* Stop tracemalloc */
```

**Review by Eclips4 — COMMENTED** [review]
Some typos

**Comment by markshannon** [comment]
@Eclips4 Thanks for spotting those.

File: Python/pylifecycle.c [hunk 8]
```diff
@@ -1136,11 +1136,12 @@ init_interp_main(PyThreadState *tstate)
             return _PyStatus_ERR("can't initialize signals");
         }
 
-        if (_PyTraceMalloc_Init(config->tracemalloc) < 0) {
-            return _PyStatus_ERR("can't initialize tracemalloc");
+        if (config->tracemalloc) {
+           if ( _PyTraceMalloc_Start(config->tracemalloc) < 0) {
```
Comment (JelleZijlstra) [inline Python/pylifecycle.c]:
```suggestion
           if (_PyTraceMalloc_Start(config->tracemalloc) < 0) {
```
(drive-by style nit)

File: Include/tracemalloc.h [hunk 9]
```diff
@@ -33,6 +33,40 @@ PyAPI_FUNC(int) PyTraceMalloc_Untrack(
...
+PyAPI_FUNC(int) _PyTraceMalloc_GetTracebackLimit(void);
+
+/* Get the memory usage of tracemalloc in bytes */
+PyAPI_FUNC(size_t) _PyTraceMalloc_GetMemory(void);
+
+/* Get the current size and peak size of traced memory blocks as a 2-tuple */
+PyAPI_FUNC(PyObject *) _PyTraceMalloc_GetTracedMemory(void);
+
+/* Set the peak size of traced memory blocks to the current size */
+PyAPI_FUNC(void) _PyTraceResetPeak(void);
```
Comment (brandtbucher) [inline Include/tracemalloc.h]: Why not `_PyTraceMalloc_ResetPeak`?

File: Python/tracemalloc.c [hunk 10]
```diff
@@ -0,0 +1,1560 @@
...
+    TABLES_LOCK();
+    size = tracemalloc_traced_memory;
+    peak_size = tracemalloc_peak_traced_memory;
+    TABLES_UNLOCK();
+
+    return Py_BuildValue("nn", size, peak_size);
+}
+
+void
+_PyTraceResetPeak(void)
```
Comment (brandtbucher) [inline Python/tracemalloc.c]: Ditto.

**Review by brandtbucher — APPROVED** [review]
Looks good (I didn't dig too closely into the moved code, but I'm trusting you that it's indeed a "dumb" refactoring).

Just one question on naming:
